### PR TITLE
Remove profiler storages

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by Tobion** [comment]
Please add the `@deprecated` phpdoc.

**Comment by fabpot** [comment]
You also need to add `@group legacy` for test files related to the deprecated classes

**Comment by fabpot** [comment]
`src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php` should also be updated to trigger a deprecation notice for people not using the filesystem.

**Comment by javiereguiluz** [comment]
Added `@deprecated` and `@group legacy`.

Deprecation of config options has been done in `Configuration.php` instead of the mentioned `FrameworkExtension.php` file. Is that correct?

**Comment by fabpot** [comment]
LGTM

**Comment by Tobion** [comment]
IMO the `dsn` config makes no sense as well (as you can see with the configs for usrname/password). A better approach would have been a `storage` config option that just accepts a service. Could we change that as well?

File: UPGRADE-2.8.md [hunk 1]
```diff
@@ -332,66 +332,81 @@ DependencyInjection
...
 -----------
 
-The `profiler:import` and `profiler:export` commands have been deprecated and
-will be removed in 3.0.
+ * The `profiler:import` and `profiler:export` commands have been deprecated and
+   will be removed in 3.0.
 
-The web development toolbar has been completely redesigned. This update has
-introduced some changes in the HTML markup of the toolbar items.
+   The web development toolbar has been completely redesigned. This update has
```
Comment (stof) [inline UPGRADE-2.8.md]: this should be a different bullet point. It is not the same change the deprecating import and export

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -264,6 +264,26 @@ private function addFragmentsSection(ArrayNodeDefinition $rootNode)
     private function addProfilerSection(ArrayNodeDefinition $rootNode)
     {
         $rootNode
+            ->beforeNormalization()
+                ->ifTrue(function ($v) { return 'file:' !== substr($v['dsn'], 0, 5); })
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: you need to check whether the ``dsn`` key exists before accessing it, as you run it on unvalidated configuration.
The alternative is to put this trigger on the ``dsn`` node itself rather than on the ``profiler`` node, as it deals with a single value anyway

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -264,6 +264,26 @@ private function addFragmentsSection(ArrayNodeDefinition $rootNode)
...
+                ->then(function ($v) {
+                    @trigger_error('The profiler.dsn configuration key must start with "file:" because all the storages except the filesystem are deprecated since version 2.8 and will be removed in 3.0.', E_USER_DEPRECATED);
+
+                    return $v;
+                })
+                ->ifTrue(function ($v) { return array_key_exists('username', $v); })
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: you cannot call ``ifTrue`` multiple times in the same ``->beforeNormalization`` (it would replace the previous condition).
You need multiple ``beforeNormalization`` here.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -264,6 +264,26 @@ private function addFragmentsSection(ArrayNodeDefinition $rootNode)
...
+                ->then(function ($v) {
+                    @trigger_error('The profiler.username configuration key is deprecated since version 2.8 and will be removed in 3.0.', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: given that you need to trigger it all the time when ``username`` is configured, and you don't need to access the other fields, it would be better to put this ``beforeNormalization`` directly on the ``username`` node (and then, you can use ``->always()`` for the condition as the processing will only normalize the value if it is configured)

File: src/Symfony/Component/HttpKernel/Profiler/BaseMemcacheProfilerStorage.php [hunk 5]
```diff
@@ -33,6 +36,8 @@
      */
     public function __construct($dsn, $username = '', $password = '', $lifetime = 86400)
     {
+        @trigger_error('The '.__CLASS__.' class is deprecated since Symfony 2.8 and will be removed in 3.0. Use FileProfilerStorage instead.', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Profiler/BaseMemcacheProfilerStorage.php]: shouldn't it be triggered at the beginning of the file instead ?

File: src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php [hunk 6]
```diff
@@ -61,16 +61,12 @@ public function testFindWorksWithInvalidDates()
...
-            $this->markTestSkipped('This test requires SQLite support in your environment');
-        }
-
         $this->tmp = tempnam(sys_get_temp_dir(), 'sf2_profiler');
         if (file_exists($this->tmp)) {
             @unlink($this->tmp);
         }
 
-        $this->storage = new SqliteProfilerStorage('sqlite:'.$this->tmp);
+        $this->storage = new FileProfilerStorage('file:'.$this->tmp);
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php]: changing the storage used in tests may be done in older branches too (avoiding the need to skip the tests in some cases, as the file storage is always available)

**Comment by stof** [comment]
status: needs work

File: src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php (hunk 6 above)
Comment (javiereguiluz) [inline src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php]: I've done all the other changes except this one. Since you are talking about different branches, I guess there is nothing to do in this pull request.

Comment (stof) [inline src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php]: Well, it means it should be done in a separate PR in older branch first rather than doing the work twice in separate PR

File: src/Symfony/Component/HttpKernel/Profiler/BaseMemcacheProfilerStorage.php (hunk 5 above)
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Profiler/BaseMemcacheProfilerStorage.php]: Indeed, that's what we do everywhere else.

Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Profiler/BaseMemcacheProfilerStorage.php]: You will have to change __CLASS__ by __NAMESPACE__ + the name of the class. I can one downside: when using a sub-class as the class name won't be the right one.

File: src/Symfony/Component/HttpKernel/Profiler/MongoDbProfilerStorage.php [hunk 7]
```diff
@@ -27,6 +31,8 @@ class MongoDbProfilerStorage implements ProfilerStorageInterface
      */
     public function __construct($dsn, $username = '', $password = '', $lifetime = 86400)
     {
+        @trigger_error('The '.__CLASS__.' class is deprecated since Symfony 2.8 and will be removed in 3.0. Use FileProfilerStorage instead.', E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Profiler/MongoDbProfilerStorage.php]: Same here

File: src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php (hunk 6 above)
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Tests/Profiler/ProfilerTest.php]: see #16009

**Comment by javiereguiluz** [comment]
I've moved deprecation notices at the beginning of the classes.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -272,9 +272,27 @@ private function addProfilerSection(ArrayNodeDefinition $rootNode)
...
                         ->booleanNode('only_exceptions')->defaultFalse()->end()
                         ->booleanNode('only_master_requests')->defaultFalse()->end()
-                        ->scalarNode('dsn')->defaultValue('file:%kernel.cache_dir%/profiler')->end()
-                        ->scalarNode('username')->defaultValue('')->end()
-                        ->scalarNode('password')->defaultValue('')->end()
+                        ->scalarNode('dsn')
+                            ->defaultValue('file:%kernel.cache_dir%/profiler')
+                            ->beforeNormalization()
+                                ->always()
+                                ->thenInvalid('The profiler.dsn configuration key must start with "file:" because all the storages except the filesystem are deprecated since version 2.8 and will be removed in 3.0.')
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: That's too strong. We should still accept other storage engines, but trigger a deprecation notice in such cases (same for the username and password).

You can see an example in the same file for `csrf_protection`.

Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Done.

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -272,9 +272,39 @@ private function addProfilerSection(ArrayNodeDefinition $rootNode)
                         ->booleanNode('collect')->defaultTrue()->end()
                         ->booleanNode('only_exceptions')->defaultFalse()->end()
                         ->booleanNode('only_master_requests')->defaultFalse()->end()
-                        ->scalarNode('dsn')->defaultValue('file:%kernel.cache_dir%/profiler')->end()
-                        ->scalarNode('username')->defaultValue('')->end()
-                        ->scalarNode('password')->defaultValue('')->end()
+                        ->scalarNode('dsn')
+                            ->defaultValue('file:%kernel.cache_dir%/profiler')
+                            ->beforeNormalization()
+                                ->always()
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: this should not be ``always``. configuring a DSN is not deprecated entirely. It should not be triggered when it starts with ``file``

Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: You are right. Changed by:

```php
->ifTrue(function ($v) { return 'file:' !== substr($v, 0, 5); })
```

File: src/Symfony/Component/HttpKernel/Profiler/PdoProfilerStorage.php [hunk 10]
```diff
@@ -11,11 +11,16 @@
 
 namespace Symfony\Component\HttpKernel\Profiler;
 
+@trigger_error('The '.__NAMESPACE__.'\PdoProfilerStorage class is deprecated since Symfony 2.8 and will be removed in 3.0. Use FileProfilerStorage instead.', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Profiler/PdoProfilerStorage.php]: this should also be triggered for child classes IMO, as they are also deprecated.

Comment (javiereguiluz) [inline src/Symfony/Component/HttpKernel/Profiler/PdoProfilerStorage.php]: Done. Thanks.

**Comment by stof** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Comment by malarzm** [comment]
I'm sorry for commenting on such old PR (especially I'm almost a year late here), but what were the reasons for deprecating and removing non-filesystem storages? I went through issues backlog/changelog/update files and only found PRs doing so without justification.

Reason I'm asking is that currently colleague of mine is working on moving us from Sf 2.8 to 3.x and I was a bit dazzled by the fact he added almost the same class as removed `MongoDbProfilerStorage` to get rid of deprecation. In our case we're using central profiler storage (and we have a web-based way to access stored profiles) which is the only way for collecting profiles from application ran on few servers (behind load balancer).

**Comment by alexgurrola** [comment]
@malarzm We've come across the same issue, and created a simple [package](https://packagist.org/packages/sitetheory/profiler-storage-bundle) to help with the transition to Symfony 3.0+.

**Comment by chadwickmeyer** [comment]
@fabpot @javiereguiluz @malarzm @alexgurrola We have the exact same issue with load balancers. Anyone who uses Symfony in an enterprise context (with more than one server) can no longer use the profiler as it's intended to be used. 

And why was this removed in the first place? There doesn't seem to be any explanation of the rational for the removal of this extremely helpful feature.

**Comment by robfrawley** [comment]
If you dug into the tickets referenced by this PR you would find https://github.com/symfony/symfony/pull/15939 which describes **why** it was done. While I don't agree with the rational, if you are looking for the explanation, it's already described there.

**Comment by javiereguiluz** [comment]
Now that we'll have Symfony Flex soon, we may rethink some past decisions, like this one, and release those features as small packages (e.g. `distributed-profiler` or `database-profiler`).
